osu.Game/Overlays/ExpandableSlider.cs

```diff
@@ -0,0 +1,128 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System;
+using osu.Framework.Allocation;
+using osu.Framework.Bindables;
+using osu.Framework.Graphics;
+using osu.Framework.Graphics.Containers;
+using osu.Framework.Graphics.UserInterface;
+using osu.Framework.Localisation;
+using osu.Game.Graphics.Sprites;
+using osu.Game.Graphics.UserInterface;
+using osuTK;
+
+namespace osu.Game.Overlays
+{
+    /// <summary>
+    /// An <see cref="IExpandable"/> implementation for the UI slider bar control.
+    /// </summary>
+    public class ExpandableSlider<T, TSlider> : CompositeDrawable, IExpandableControl, IHasCurrentValue<T>
+        where T : struct, IEquatable<T>, IComparable<T>, IConvertible
+        where TSlider : OsuSliderBar<T>, new()
```

Comment:
not sure how to put this nicely, but - when you [suggested "separate sets of controls"](https://github.com/ppy/osu/pull/16574#issuecomment-1019587153), i clearly said [that would be an overcomplication](https://github.com/ppy/osu/pull/16574#issuecomment-1019589516). but this is one of several expandable controls. so i'm not sure whether my comment was disregarded or "separate sets of controls" meant something else (i shudder to even think what it would have meant, exactly).

to be more clear, below is a list of classes introduced in this diff which to me should not exist:

- `ExpandableSlider`
- `ExpandingButtonContainer`
- `IExpandable`
- `IExpandableControl`
- `IExpandingContainer`

i claim that all that is required to make this work is an `ExpandingControlContainer` with proper composition support. unless you can provide a *good* reason for introducing all of this complexity i don't see myself ever agreeing with any of this.

Reply:
I figured that would be the case, but I wanted to first push the changes I had in mind to properly discuss on the direction forward.

Going for composition seems like another way but I can't imagine it working with the way the `Editor`'s usages are, which go in a similar pattern to:
```
 - `ExpandingControlContainer<OsuButton>`
   - `SettingsToolboxGroup`
     - `DrawableRadioButton`
     - `DrawableRadioButton`
     - ...
   - `SettingsToolboxGroup`
     - `DrawableRadioButton`
     - `DrawableRadioButton`
     - ...
 - `ExpandingControlContainer<IExpandableControl> (newly added distance spacing stuff)
   - `SettingsToolboxGroup`
     - `ExpandableSlider`
```

The only way I can see it myself with full composition and no DI in children components is to use `ChildrenOfType`, and I considered that myself, but I don't believe that's a good solution at all, for the fact that `ChildrenOfType` is only supposed to be for tests and goes against the natural pattern of DI (with the sole exception of using it for retrieving whether controls nested in the hierarchy are hovered, which I myself am not sure how to potentially handle without).

That's why I went with a hierarchial structure instead along with the *unfortunate* complexity it provides, although I'm not really 100% confident about it either, but I didn't want to spend too much time on it just for a [simple feature addition](https://github.com/ppy/osu/pull/16576) neither did I want to bring it up with a bad UX.

Reply:
i personally would rather see `.ChildrenOfType<>()` used for hover detection (albeit used maybe in a bit of a smarter manner than just calling it every `Update()`, e.g. combined with a `Cached` or something) than whatever this is. you say that you "didn't want to spend too much time on it" but such a solution would be vastly shorter and therefore easier to write *and* review.

Reply:
Well, I thought the exact opposite and considered `.ChildrenOfType<>()` to be a bad solution that would be best not to use when avoidable. I'm fine with proceeding on that direction, but curious to hear @peppy's opinion first.